db/db_iter.cc

```diff
@@ -611,21 +613,27 @@ inline bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check)
             !iter_.iter()->IsKeyPinned() || !pin_thru_lifetime_ /* copy */);
         skipping = false;
         num_skipped = 0;
+        reseek_done = false;
       }
     }
 
     // If we have sequentially iterated via numerous equal keys, then it's
     // better to seek so that we can avoid too many key comparisons.
-    if (num_skipped > max_skip_ && CanReseekToSkip()) {
+    //
+    // Do not reseek if we have already attempted to reseek previously to avoid
+    // infinite loops.
+    if (num_skipped > max_skip_ && !reseek_done) {
       is_key_seqnum_zero_ = false;
       num_skipped = 0;
+      reseek_done = true;
       std::string last_key;
       if (skipping) {
         // We're looking for the next user-key but all we see are the same
         // user-key with decreasing sequence numbers. Fast forward to
         // sequence number 0 and type deletion (the smallest type).
-        AppendInternalKey(&last_key, ParsedInternalKey(saved_key_.GetUserKey(),
-                                                       0, kTypeDeletion));
+        AppendInternalKey(&last_key,
+                          ParsedInternalKey(saved_key_.GetUserKey(), 0,
+                                            kValueTypeForSeekForPrev));
```

Comment:
This change isn't essentially and I can revert this, but I just thought using `kValueTypeForSeekForPrev` conveyed intent better since we just want the "smallest" type, and `kTypeDeletion` just happens to be the smallest marker type. It seems like we use `kValueTypeForSeekForPrev` elsewhere in this file when we want to seek to the smallest type.

Reply:
Iterator's code is overly complicated and error-prone. When we have to touch it, it is safer to keep the changes to a minimum essentials.